### PR TITLE
Implemented ZC_SKILLINFO_UPDATE2 packet

### PR DESCRIPTION
**Pull request by AoShinRO**
* **Addressed Issue(s)**: 

* **Server Mode**: both

* **Description of Pull Request**: Implemented ZC_SKILLINFO_UPDATE2 usage

### CODE REVIEW

File: src/map/clif.cpp [hunk 1]
```diff
@@ -5756,28 +5756,37 @@ void clif_skillup(map_session_data *sd, uint16 skill_id, int lv, int range, int
 
 /// Updates a skill in the skill tree (ZC_SKILLINFO_UPDATE2).
```
Comment (AoShinRO) [inline src/map/clif.cpp:5757]:
```suggestion
/// Updates a skill in the skill tree
```

File: src/map/clif.cpp [hunk 2]
```diff
@@ -5756,28 +5756,37 @@ void clif_skillup(map_session_data *sd, uint16 skill_id, int lv, int range, int
...
 /// 07e1 <skill id>.W <type>.L <level>.W <sp cost>.W <attack range>.W <upgradable>.B
```
Comment (AoShinRO) [inline src/map/clif.cpp:5758]:
```suggestion
/// 07e1 <skill id>.W <type>.L <level>.W <sp cost>.W <attack range>.W <upgradable>.B (ZC_SKILLINFO_UPDATE2)
```

File: src/map/clif.cpp [hunk 3]
```diff
@@ -5756,28 +5756,37 @@ void clif_skillup(map_session_data *sd, uint16 skill_id, int lv, int range, int
...
-void clif_skillinfo(map_session_data *sd,int skill_id, int inf)
-{
-	nullpo_retv(sd);
-
-	const int fd = sd->fd;
+/// 0b33 <skill id>.W <type>.L <level>.W <sp cost>.W <attack range>.W <upgradable>.B <level2>.W
```
Comment (AoShinRO) [inline src/map/clif.cpp:5759]:
```suggestion
/// 0b33 <skill id>.W <type>.L <level>.W <sp cost>.W <attack range>.W <upgradable>.B <level2>.W (ZC_SKILLINFO_UPDATE3)
```

File: src/map/clif.cpp [hunk 4]
```diff
@@ -5754,30 +5754,39 @@ void clif_skillup(map_session_data *sd, uint16 skill_id, int lv, int range, int
...
+	p.sp = skill_get_sp(skill_id,sd.status.skill[idx].lv);
+	p.range2 = skill_get_range2(&sd.bl,skill_id,sd.status.skill[idx].lv,false);
+	p.upFlag = 0;
+
+	if(inf >= 0)
+		p.inf = inf;
 	else
-		WFIFOB(fd,14) = 0;
-	WFIFOSET(fd,packet_len(0x7e1));
+		p.inf = skill_get_inf(skill_id);
```
Comment (Atemo) [inline src/map/clif.cpp:5779]:
```suggestion
	p.inf = skill_get_inf(skill_id);
```

File: src/map/clif.cpp [hunk 5]
```diff
@@ -5754,30 +5754,39 @@ void clif_skillup(map_session_data *sd, uint16 skill_id, int lv, int range, int
...
+
+	if( sd.status.skill[idx].flag == SKILL_FLAG_PERMANENT && sd.status.skill[idx].lv < skill_tree_get_max(skill_id, sd.status.class_) )
+		p.upFlag = 1;
```
Comment (Atemo) [inline src/map/clif.cpp:5782]:
```suggestion
		p.upFlag = 1;
	else
		p.upFlag = 0;
```

File: src/map/clif.cpp [hunk 6]
```diff
@@ -5754,30 +5754,39 @@ void clif_skillup(map_session_data *sd, uint16 skill_id, int lv, int range, int
...
-/// 07e1 <skill id>.W <type>.L <level>.W <sp cost>.W <attack range>.W <upgradable>.B
-void clif_skillinfo(map_session_data *sd,int skill_id, int inf)
-{
-	nullpo_retv(sd);
-
-	const int fd = sd->fd;
+/// Updates a skill in the skill tree
+/// 07e1 <skill id>.W <type>.L <level>.W <sp cost>.W <attack range>.W <upgradable>.B (ZC_SKILLINFO_UPDATE2)
+/// 0b33 <skill id>.W <type>.L <level>.W <sp cost>.W <attack range>.W <upgradable>.B <level2>.W (ZC_SKILLINFO_UPDATE3)
+void clif_skillinfo(map_session_data& sd,int skill_id, int inf){
```
Comment (Atemo) [inline src/map/clif.cpp:5760]:
```suggestion
void clif_skillinfo( map_session_data& sd, int32 skill_id ){
```

File: src/map/clif.hpp [hunk 7]
```diff
@@ -930,7 +930,7 @@ void clif_class_change_target(struct block_list *bl,int class_, int type, enum s
 
 void clif_skillinfoblock(map_session_data *sd);
 void clif_skillup(map_session_data *sd, uint16 skill_id, int lv, int range, int upgradable);
-void clif_skillinfo(map_session_data *sd,int skill_id, int inf);
+void clif_skillinfo(map_session_data &sd,int skill_id, int inf = -1);
```
Comment (Atemo) [inline src/map/clif.hpp:933]:
```suggestion
void clif_skillinfo( map_session_data &sd, int32 skill_id );
```

File: src/map/clif.cpp [hunk 8]
```diff
@@ -5754,30 +5754,39 @@ void clif_skillup(map_session_data *sd, uint16 skill_id, int lv, int range, int
...
+
+	p.packetType = HEADER_ZC_SKILLINFO_UPDATE2;	
+	p.id = skill_id;
+	p.level = sd.status.skill[idx].lv;
+	p.sp = skill_get_sp(skill_id,sd.status.skill[idx].lv);
+	p.range2 = skill_get_range2(&sd.bl,skill_id,sd.status.skill[idx].lv,false);
+	p.upFlag = 0;
+
+	if(inf >= 0)
+		p.inf = inf;
```
Comment (Atemo) [inline src/map/clif.cpp:5777]: upFlag is defined after
'inf' is always 0 in our code
```suggestion
	p.sp = static_cast<uint16>( skill_get_sp(skill_id,sd.status.skill[idx].lv) );
	p.range2 = static_cast<uint16>( skill_get_range2(&sd.bl,skill_id,sd.status.skill[idx].lv,false) );
```

File: src/map/skill.cpp [hunk 9]
```diff
@@ -3231,19 +3231,21 @@ void skill_combo_toggle_inf(struct block_list* bl, uint16 skill_id, int inf){
 		case MO_COMBOFINISH:
 		case CH_TIGERFIST:
 		case CH_CHAINCRUSH:
-			if (sd) clif_skillinfo(sd,MO_EXTREMITYFIST, inf);
+			if (sd != nullptr)
+				clif_skillinfo(*sd,MO_EXTREMITYFIST, inf);
```
Comment (Atemo) [inline src/map/skill.cpp:3235]: 'inf' from skill_combo_toggle_inf is always 0
```suggestion
				clif_skillinfo(*sd,MO_EXTREMITYFIST);
```

File: src/map/skill.cpp [hunk 10]
```diff
@@ -3231,19 +3231,21 @@ void skill_combo_toggle_inf(struct block_list* bl, uint16 skill_id, int inf){
...
 			break;
 		case TK_JUMPKICK:
-			if (sd) clif_skillinfo(sd,TK_JUMPKICK, inf);
+			if (sd != nullptr)
+				clif_skillinfo(*sd,TK_JUMPKICK, inf);
```
Comment (Atemo) [inline src/map/skill.cpp:3239]:
```suggestion
				clif_skillinfo(*sd,TK_JUMPKICK);
```

File: src/map/skill.cpp [hunk 11]
```diff
@@ -3231,19 +3231,21 @@ void skill_combo_toggle_inf(struct block_list* bl, uint16 skill_id, int inf){
...
 			break;
 		case MO_TRIPLEATTACK:
-			if (sd && pc_checkskill(sd, SR_DRAGONCOMBO) > 0)
-				clif_skillinfo(sd,SR_DRAGONCOMBO, inf);
+			if (sd != nullptr && pc_checkskill(sd, SR_DRAGONCOMBO) > 0)
+				clif_skillinfo(*sd,SR_DRAGONCOMBO, inf);
```
Comment (Atemo) [inline src/map/skill.cpp:3243]:
```suggestion
				clif_skillinfo(*sd,SR_DRAGONCOMBO);
```

File: src/map/skill.cpp [hunk 12]
```diff
@@ -3231,19 +3231,21 @@ void skill_combo_toggle_inf(struct block_list* bl, uint16 skill_id, int inf){
...
 			break;
 		case SR_FALLENEMPIRE:
-			if (sd){
-				clif_skillinfo(sd,SR_GATEOFHELL, inf);
-				clif_skillinfo(sd,SR_TIGERCANNON, inf);
+			if (sd != nullptr){
+				clif_skillinfo(*sd,SR_GATEOFHELL, inf);
+				clif_skillinfo(*sd,SR_TIGERCANNON, inf);
```
Comment (Atemo) [inline src/map/skill.cpp:3248]:
```suggestion
				clif_skillinfo(*sd,SR_GATEOFHELL);
				clif_skillinfo(*sd,SR_TIGERCANNON);
```

**Review by aleos89 — APPROVED** [review]

**Review by Atemo — APPROVED** [review]

**Pull request closed without merging**